efRefactor Case Importer

esoergel opened the pull request:
I'd definitely go commit-by-commit, which should be pretty straightforward to follow.  

My main goal here was to make this code more comprehensible.  I did find a number bugs and odd quirks in behavior in the process, which I addressed.  It should be clear when behavior is intentionally changed.

### Fixes

* Make `too_many_matches` a normal error like any other, so it is more meaningful to the user.  It'll also show which rows have too many matches
* Add error message when parent lookup by external ID fails
* Fix flawed logic that would sometimes result in cases referencing parents that hadn't yet been uploaded (which would fail)
* Log unknown failures (we're currently silencing them, so some bugs are really hard to understand)
* Fix weird "-1 rows were matched and updated" thing
* Keep track of status per row.   This used to log rows to each error category, and I inverted that.  This means the task results should be much more reliable.  Currently, you see this:
![image](https://user-images.githubusercontent.com/2367539/57950910-9888e480-78b6-11e9-8416-72346ab5508a.png)
Those rows are all logged as successful, then they later fail during the "commit" stage, but we don't un-mark them successful, so you get contradictory messages.

@kaapstorm

esoergel committed:
Remove unused vars

esoergel committed:
Split do_import into its own file

esoergel committed:
Move do_import to a class

esoergel committed:
Patch global var rather than passing it all around

esoergel committed:
Move import params to __init__

esoergel committed:
Split out submit_caseblocks

esoergel committed:
Move individual row handling to its own method

esoergel committed:
Extract caseblock chunking

esoergel committed:
Use exceptions for short-circuiting errors

esoergel committed:
Pull out most chunks of handle_row

esoergel committed:
Simplify row_number handling

use a multi-letter variable name, start at 1, and don't pass it around
as frequently

esoergel committed:
Remove flawed logic

This was mistaken in several ways:

* It only happened on the final form submission - errors in earlier chunks
  are exempt.
* It deducts 1 from match_count - there's no reason to believe there is
  precisely one failed case
* Why match_count?  Couldn't the error have been with a new case
  (created_count)
* It also added a chunk whether or not there were any cases in that
  final submission

I do think this is related to the issue which triggered this refactor,
but I'll look in to a proper fix later

esoergel committed:
Use clearer names

esoergel committed:
Log errors also when getting parent by external id

lookup_case already handles lookup by id or external_id, so this
consolidates the two calls and logs the error in both cases

esoergel committed:
Always reset uncreated_external_ids

regardless of how they're created

esoergel committed:
Notify of exception before ignoring

What happens here?  Is this something we should be addressing?

esoergel committed:
Raise error on appropriate column

esoergel commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -240,28 +240,22 @@ def set_owner_id(self, name_cache, id_cache):
+                self.parent_ref: (parent_case.type, parent_case.case_id)
+            }
+        else:
+            raise exceptions.InvalidParentId('parent_id')
```
https://github.com/dimagi/commcare-hq/pull/24352/commits/755a6617c0949ae6b2ff1c174b0cc80a160e1193

esoergel committed:
Make too_many_matches a normal error

esoergel committed:
lookup_case filters by case type

So this check is redundant

esoergel committed:
Consolidate external_id logic

esoergel committed:
Store results for every row

This is especially useful for instances where a row is initially marked
as successful, but then the subsequent form submission fails.
Currently, such rows count both towards successes and failures.  This
will make the results display more accurate.

esoergel committed:
Treat actual error as such

If there's a form error here, I _think_ it's a back-end problem, not a
user problem.  Figure it makes more sense to raise than to swallow.
(plus `form.problem` is not a row number)

esoergel committed:
lint

esoergel commented on corehq/apps/case_importer/util.py:
```diff
@@ -170,67 +171,48 @@ def convert_custom_fields_to_struct(config):
+        self._results[row_num] = self.UPDATED
+
+    def to_json(self):
+        counts = Counter(self._results.values())
```
Ugh, you are such a stickler - I _am_ iterating.

proteusvacuum commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -0,0 +1,278 @@
+            case_type=self.config.case_type,
+            case_name=self.case_name or '',
+            update=self.fields_to_update,
+            **self.extras
```
building on my comment from before - I think there might be a way that you can just generate these when you call `self.extras`, so it would be `self.get_extras()` which would then return all of the appropriate items so that the `_Importer` class doesn't have to populate them.

proteusvacuum reviewed (commented):
Fun!!

esoergel commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -25,248 +25,262 @@
+        self.num_chunks += 1
+
+    @property
+    def outcome(self):
```
Do you mean like whether or not _any_ rows completed successfully?  Yes, I do think that needs to be addressed, as the way it's handled now is often wrong.  This would be the right place for that to live, I think.

proteusvacuum replied:
That was a bit cryptic, sorry.
The way this `_Importer` class is set up, the `_Importer(...).outcome()` will be meaningless unless you've already called `_Importer(...).do_import()`. I was suggesting that you explicitly say that somehow, or just make the call to `do_import` from this `outcome()` method. That way this class only has a single publicly accessible method and everything else is hidden tidily away.

kaapstorm replied:
I do like the idea of fewer public methods, but the name "outcome()" doesn't make me think that calling it will also do an import (although that could just be me). For fewer methods, tidying everything else away, I'd lean towards renaming "outcome()" to "_outcome()" and having "do_import()" return it.

esoergel replied:
This was changed around a bit in a subsequent commit - here's how the main function works now:

https://github.com/dimagi/commcare-hq/blob/3f2439c95be28188e417ccb95e8a5888a6f8bfa6/corehq/apps/case_importer/do_import.py#L38-L41

It still has essentially the same pitfalls as y'all are describing, though.  Let me explain my reasoning a little, I think we're all on roughly the same page.

This whole file can be summed up by that simple interface above - a function which accepts a handful of arguments and returns a simple JSON representation of the result.  However, far too much happens to fit that in a single function (hence this PR).

Given the amount of shared state, I think hosting it on a class was the most straightforward way to break it up.  This presents problems, as we really only want that simple function interface, but classes aren't supposed to do work in the constructor, and if you have fewer than two public methods, you don't need a class.  What I settled on was using a private class coupled tightly to the `do_import` function, which knows how to drive it.  This gives the external interface I want, and makes it more acceptable to bend some rules.  I do like returning the result directly from `do_import` though, I'll make that change.

Does that sound like the best way to approach this situation?  Simple wrapper function that instantiates the class and calls the `run()` method, with everything else hidden inside the class?

esoergel replied:
Another option is this:
```python
class do_import(object):
    def __init__(self, spreadsheet, config, domain, task=None, record_form_callback=None):
        ...
        for row in spreadsheet:
            self.import_row
        return self.results.to_json()
```
but that just feels so wrong.

kaapstorm replied:
> but that just feels so wrong.

**\*shudder\*** ... yes

> if you have fewer than two public methods, you don't need a class.

:thinking: Because Uncle Bob says so, because convention, or because it has been shown that you can write the same code functionally and it will be simpler? (The last one does sound possible, but I can't see whether it would always be true.)

proteusvacuum replied:
>  I do like returning the result directly from do_import though, I'll make that change.

I like this option the best, and is kind of the pattern I've been following when doing single function -> class refactors. 

> if you have fewer than two public methods, you don't need a class

Says who? I think you can even have _no_ public methods and still have a useful, easy to read class that acts as a data wrapper.

esoergel replied:
There's a pylint warning `too-few-public-methods`.  I agree with it in principle, with the exception of things like dataclasses.  I do think in this case the one master method approach yields the best result.  I could even see adopting some kind of naming convention around this to make it clearer that this is a class that acts as a single function, though I don't wanna walk down that road right now.

esoergel commented on corehq/apps/case_importer/exceptions.py:
```diff
@@ -35,3 +36,87 @@ class ImporterExcelFileEncrypted(ImporterExcelError):
+
+
+class DuplicateLocationName(CaseRowError):
+    title = ugettext_noop('Duplicated Location Name')
```
The titles are explicitly translated here:
https://github.com/dimagi/commcare-hq/pull/24352/files#diff-d3d016b9ef00a1b62fc1fba9d4c6766cL220
I did try to switch that to lazy at some point, but hit an error about using a proxy object in some context - don't recall what it was.  Probably resolvable if that feels important, though.

proteusvacuum replied:
Oh, I missed that. I think that's fine, but might be worth putting a note about that somewhere. Most places I see `noop` I think were done in the dark ages when we didn't know the difference.

esoergel commented on corehq/apps/case_importer/migrations/0013_make_duplicates_errors.py:
```diff
@@ -0,0 +1,46 @@
+
+
+def migrate_record_duplicates(apps, schema_editor):
+    for record in iterator(CaseUploadRecord.objects.all()):
```
excellent point, will update, thanks

esoergel commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -33,10 +33,7 @@ def __init__(self, domain, config, task, record_form_callback):
-        self.match_count = 0
-        self.num_chunks = 0
-        self.errors = importer_util.ImportErrorDetail()
+        self.results = importer_util.ImportResults()
```
:bear: 
Yeah, I like that - some of it is referenced externally (there's a custom version of the case importer that re-uses a lot of this stuff), but I suppose it's better to have a real home than stuffing it all in `util`

esoergel commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -0,0 +1,278 @@
+        row.set_owner_id(self.name_cache, self.id_cache)
+        row.set_parent_id(self.log_case_lookup)
+        row.set_date_opened()
+        row.set_external_id(is_new_case=not case)
```
I wrote a comment about this exact block, actually, but forgot to submit it.  This is one of the remaining pieces that stands out as ugly to me too.  What I'd really like to do with this is pass in everything `CaseImportRow` needs up-front, then thereafter only access getter methods, without changing state internal to `CaseImportRow`.  Unfortunately that'd require passing in like 12 parameters to the constructor, which is also gross.  

For now I kept control flow here, where all the necessary objects are.  Can you think of a good way to move control flow to the other class without giving it access to essentially all the attributes of this parent class?

One possible approach I considered would be to group all these dependencies,  like `self.results` - say, a object that handles caching, and a light wrapper around metadata stuff like domain, user, and config.  That doesn't really reduce the coupling between the classes, but it would compartmentalize it a bit.

esoergel replied:
Just saw your other comment:
> building on my comment from before - I think there might be a way that you can just generate these when you call self.extras, so it would be self.get_extras() which would then return all of the appropriate items so that the _Importer class doesn't have to populate them.

I definitely agree.  Do you think it'd be preferable to pass in everything up front and avoid mutating the class?  That's kind of where I'm leaning now, though I would like to somehow minimize the dependencies between the two classes.

kaapstorm replied:
I'd second the approach to avoid mutating the class. I'd argue that it makes it more testable and easier to reason about, even if passing stuff in up-front could be ugly.

proteusvacuum replied:
I think both of the caches can be extracted out to use quickcache so you don't have to pass the dicts around. 
In general though I also think it's preferable to avoid mutating the class from outside.

esoergel replied:
> I think both of the caches can be extracted out to use quickcache so you don't have to pass the dicts around.

I'm not sure - right now it's an in-memory lookup, and switching to quickcache would mean using redis or memcached.  Tying the cache to this code means it can persist for as short or as long as the case import takes (which can vary considerably), and the lookups are always in memory.

I also considered making a object to store these caches and the associated code, say `CaseOwnerAccessor` or something.  I do think that'd be a good way to further simplify this code, although I'm gonna hold off on that, as I'm trying to somewhat limit the scope of this refactor.

esoergel commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -0,0 +1,278 @@
+        self.domain = domain
+        self.config = config
+        self.task = task
+        self._record_form_callback = record_form_callback
```
I don't follow - do you mean that it'll always be set, and I can just call it directly?  Or that it could be refactored out into the task and completely out of this file?

kaapstorm commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -0,0 +1,272 @@
+            else:
+                if record_form_callback:
+                    record_form_callback(form.form_id)
+                properties = set().union(*[set(c.dynamic_case_properties().keys()) for c in cases])
```
... or ...
```python
    properties = {p for c in cases for p in c.dynamic_case_properties().keys()}
```
... although it's debatable what's more readable, so maybe just consider this a totally optional nit.

proteusvacuum replied:
^ That is more readable, and is a bit more :snake:

kaapstorm replied:
Oh, oh, for more :snake: we can drop `keys()`:
```python
    properties = {p for c in cases for p in c.dynamic_case_properties()}
```
:grin:

esoergel replied:
:+1: 
Although I actually don't like the implicit keys thing - makes me forget that I'm dealing with a dict.

kaapstorm reviewed (approved):
This is excellent!

esoergel committed:
PR feedback

esoergel commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -67,13 +67,7 @@ def do_import(self, spreadsheet):
-        try:
-            self._submit_caseblocks(self._caseblocks)
-        except exceptions.CaseRowError:
-            self.match_count -= 1
```
FWIW, I did end up implementing that logic: https://github.com/dimagi/commcare-hq/pull/24352/commits/84090915c84563ec11183dbd0092ebfdbaa87f65

esoergel committed:
Fix migration

esoergel committed:
Move single-use fns to do_import and mark private

esoergel committed:
Return results directly from do_import

esoergel committed:
PR feedback

proteusvacuum commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -0,0 +1,487 @@
+    group, then location
+    '''
+    if name in cache:
+        return cache[name]
```
Can you just use quickcache for this?

proteusvacuum commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -0,0 +1,487 @@
+
+def _is_valid_id(uploaded_id, domain, cache):
+    if uploaded_id in cache:
+        return cache[uploaded_id]
```
could use quickcache for this too

esoergel committed:
Don't mutate _CaseImportRow

esoergel committed:
Minor code cleanup

esoergel commented:
Could a get a :heavy_check_mark: on the last couple changes?  I switched `_CaseImportRow` to be stateless.

kaapstorm reviewed (approved):
:+1: Nice! I like what you've done with `existing_case()`, and I don't think 7 params is so bad.

proteusvacuum reviewed (approved):
super duper :tada:

esoergel commented:
Thanks for the design feedback and thorough review, @proteusvacuum and @kaapstorm !

esoergel committed:
rm trailing comma

kaapstorm commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -277,7 +277,7 @@ def get_create_caseblock(self):
         extras = self._get_caseblock_kwargs()
         if self.uploaded_owner_id or self.uploaded_owner_name:
-            extras['owner_id'] = self._get_owner_id(),
+            extras['owner_id'] = self._get_owner_id()
```
That can be hard to spot. Is there a way to get Stickler to insist that we write tuples with brackets? (A quick google revealed nothing. If you aren't immediately sure, and when I have some spare time, I'll dig a little deeper. It's just that this is the second issue in a week related to a single comma.)

esoergel replied:
Agreed!  I can't think of a situation where it's a good idea to declare a tuple that way, nor of a time when that was someone's intent.  Looks like pylint does have a checker called `trailing-comma-tuple`, though I think it's python3 only:

![image](https://user-images.githubusercontent.com/2367539/58339414-932c1c80-7e17-11e9-9b84-ed007a928e09.png)

(I've been using python2 still, so I didn't see it)

esoergel committed:
Fix off-by-one error

This introduces a race condition.  The celery task returns a task status
with a result detailing the outcome of the upload, which is fetched and
saved immediately once the task is marked as complete, either by the
polling request on the case importer page or by a separate save_result
task kicked off at the end of the import task.  With this off-by-one
error, the task result briefly has `None` in place of the outcome JSON,
which is then saved to the upload.  Introducing a delay fixes that.

I'm not sure why fixing this progress info resolves that issue, but it
does.  Perhaps something along the way sees progress>total and declares
the task done premptively?

esoergel commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -56,7 +56,7 @@ def __init__(self, domain, config, task, record_form_callback):
     def do_import(self, spreadsheet):
         for row_num, row in enumerate(spreadsheet.iter_row_dicts(), start=1):
-            set_task_progress(self.task, row_num, spreadsheet.max_row)
+            set_task_progress(self.task, row_num - 1, spreadsheet.max_row)
```
Even spookier - this bug only triggers the race condition when my branch is merged with master.  :man_shrugging:
